pkg/frontend/querymiddleware/cardinality.go

```diff
@@ -0,0 +1,172 @@
+	}
+
+	var res Response
+	if withCardinalityEstimate != nil {
+		res, err = c.next.Do(ctx, withCardinalityEstimate)
+	} else {
+		res, err = c.next.Do(ctx, request)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	statistics := stats.FromContext(ctx)
+	if statistics == nil {
+		return res, nil
+	}
+
+	actualCardinality := statistics.GetFetchedSeriesCount()
+	if !isCardinalitySimilar(actualCardinality, estimatedCardinality) {
+		c.storeCardinalityForKey(ctx, k, actualCardinality)
+	}
+
+	c.maintainStatistics(estimatedCardinality, actualCardinality, statistics, spanLog)
+
+	return res, nil
+}
+
+// lookupCardinalityForKey fetches a cardinality estimate for the given key from
+// the results cache.
+func (c *cardinalityEstimation) lookupCardinalityForKey(ctx context.Context, key string) (uint64, bool) {
+	if c.cache == nil {
+		return 0, false
+	}
+	cacheKey := cacheHashKey(key)
+	res := c.cache.Fetch(ctx, []string{cacheKey})
+	if val, ok := res[cacheKey]; ok {
+		cardinality := &QueryCardinality{}
+		err := proto.Unmarshal(val, cardinality)
+		if err != nil {
+			level.Warn(c.logger).Log("msg", "failed to unmarshal cardinality estimate")
+			return 0, false
+		}
+		return cardinality.Estimated, ok
+	}
+	return 0, false
+}
+
+// storeCardinalityForKey stores a cardinality estimate for the given key in the
+// results cache.
+func (c *cardinalityEstimation) storeCardinalityForKey(ctx context.Context, key string, count uint64) {
+	if c.cache == nil {
+		return
+	}
+	m := &QueryCardinality{Estimated: count}
+	marshaled, err := proto.Marshal(m)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "failed to marshal cardinality estimate")
+	}
+	// The store is executed asynchronously, potential errors are logged and not
+	// propagated back up the stack.
+	c.cache.Store(ctx, map[string][]byte{cacheHashKey(key): marshaled}, cardinalityEstimateTTL)
+}
+
+func (c *cardinalityEstimation) maintainStatistics(estimate uint64, actual uint64, s *stats.Stats, span *spanlogger.SpanLogger) {
+	if estimate > 0 {
+		c.estimationError.Observe(math.Abs(float64(actual) - float64(estimate)))
+		s.FetchedSeriesEstimate = estimate
+		span.LogKV("estimated cardinality", estimate, "actual cardinality", actual)
+	}
+}
+
+func isCardinalitySimilar(actualCardinality, estimatedCardinality uint64) bool {
+	estimate := float64(estimatedCardinality)
+	actual := float64(actualCardinality)
+	return estimate > (1-cacheErrorToleranceFraction)*actual && estimate < (1+cacheErrorToleranceFraction)*actual
+}
+
+// generateCacheKey generates a key to cache a request's cardinality estimate under.
+func generateCacheKey(userID string, r Request, bucketSize time.Duration) string {
+	startBucket := r.GetStart() / bucketSize.Milliseconds()
+	rangeBucket := (r.GetEnd() - r.GetStart()) / bucketSize.Milliseconds()
+
+	if rangeBucket == 0 {
+		return fmt.Sprintf("%s:%s:%d", userID, r.GetQuery(), startBucket)
+	}
+	return fmt.Sprintf("%s:%s:%d:%d", userID, r.GetQuery(), startBucket, rangeBucket)
```

Comment:
[design] Better to add a well known short prefix to cache keys (e.g. we do the same in the store-gateway). You may have "QS" for "query statistics".

```suggestion
	return fmt.Sprintf("QS:%s:%s:%d:%d", userID, r.GetQuery(), startBucket, rangeBucket)
```